Support graphql file upload

chalermpong opened the pull request:
Hi there
This pr provides capability to upload file (#640) using graphql according to this spec: https://github.com/jaydenseric/graphql-multipart-request-spec.

chalermpong commented:
@sav007 @digitalbuddha Could you help suggest me how can I fix 2 failed checking. I try to fix them but really no idea. 
- CI failed: The unit test runs successfully on my machine but failed on Travis CI
- Deploy preview failed: Error: watch /opt/build/repo/docs/gatsby-config.js ENOSPC

sav007 commented:
Seems like integration tests failing on CI. The second don't bother as not that not really relevant, will ask some folks from Apollo to take a look.
Meantime I will take a closer look at your PR.

sav007 commented:
I can't find the file `SingleUploadMutation.java`, is it missing in PR?

sav007 commented on apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java:
```diff
@@ -175,4 +182,67 @@ static ByteString httpRequestBody(Operation operation, ScalarTypeAdapters scalar
+  static RequestBody transformToMultiPartIfUploadExists(RequestBody originalBody, Operation operation)
+      throws IOException {
+    //ArrayList<GraphqlUpload> allUploads = new ArrayList<>();
+    HashMap<String, String[]> filesMap = new HashMap<>();
```
Not sure why second type is an array `HashMap<String, String[]>`?
I think that you can instead of 2 maps have this structure: `List<FileUpload>` where:
```
FileUpload(key:String, mimeType:String, file:File)
```

`key`  = `"variables." + variableName`
`mimeType` = `GraphqlUpload.mimeType`
`file` = `GraphqlUpload.file`

The order in the list is the index in multipart body.
What do you think?

chalermpong replied:
You're right. I will revise the code.

sav007 commented on apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java:
```diff
@@ -175,4 +182,67 @@ static ByteString httpRequestBody(Operation operation, ScalarTypeAdapters scalar
+      HashMap<String, GraphqlUpload> uploads) throws IOException {
+    Buffer buffer = new Buffer();
+    JsonWriter jsonWriter = JsonWriter.of(buffer);
+    jsonWriter.setSerializeNulls(true);
```
Don't think you need this, as there are no null values?

sav007 commented on apollo-api/src/main/java/com/apollographql/apollo/api/GraphqlUpload.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.io.File;
+
+public class GraphqlUpload {
```
Should we call it `GraphQLUpload `?

chalermpong replied:
Is just `Upload` or `FileUpload` ok? any suggestion...

sav007 commented on README.md:
````diff
@@ -240,6 +240,28 @@ apollo {
+Then create your mutation with mimetype and a valid `File`.
+
+```java
+  mutationSingle = SingleUploadMutation.builder()
````
I think it's better to provide the GraphQL query itself in this example:
```
mutation SingleUpload($file: Upload!) {
  singleUpload(file: $file) {
    id
    path
    filename
    mimetype
  }
}
```

sav007 replied:
And mention that GraphQL schema should have some custom scalar type named for example `Upload `

chalermpong commented:
@sav007 Thank you for your comment. I revised my code as your comment and rebased my pr. Please have a look. 
I also fixed compile error in ApolloServerInterceptorTest so that I can run the unit test.

sav007 commented on apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java:
```diff
@@ -240,4 +248,70 @@ static void addExtensionsUrlQueryParameter(HttpUrl.Builder urlBuilder, Operation
+  private static class FileUploadMeta {
+    public String key;
+    public String mimetype;
+    public File file;
```
small nit: can be final.

chalermpong replied:
Fixed

sav007 commented on apollo-api/src/main/java/com/apollographql/apollo/api/FileUpload.java:
```diff
@@ -0,0 +1,13 @@
+
+public class FileUpload {
+    public String mimetype;
+    public File file;
```
class and fields should be final.

chalermpong replied:
Fixed

chalermpong commented on apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java:
```diff
@@ -240,4 +248,70 @@ static void addExtensionsUrlQueryParameter(HttpUrl.Builder urlBuilder, Operation
+          && ((Collection<Object>) value).iterator().next() instanceof FileUpload) {
+        Collection<FileUpload> uploads = (Collection<FileUpload>) value;
+        int varFileIndex = 0;
+        for (FileUpload upload: uploads) {
```
Fixed

chalermpong commented on apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/ApolloServerInterceptor.java:
```diff
@@ -240,4 +248,70 @@ static void addExtensionsUrlQueryParameter(HttpUrl.Builder urlBuilder, Operation
+    JsonWriter jsonWriter = JsonWriter.of(buffer);
+    jsonWriter.beginObject();
+    for (int i = 0; i < fileUploads.size(); i++) {
+      jsonWriter.name("" + i).beginArray();
```
Fixed

sav007 commented:
Could you pls comment this line in `travis.yml`:
```
# - .buildscript/integration_tests_composite.sh
```

So that the build will pass, as composite build requires to changes to api module published first to maven.

chalermpong commented:
Commented the line in `travis.yml`

nikname commented:
Any idea about when this feature will be released? Thanks in advice

digitalbuddha commented:
The snapshot should be live, Ideally we'd wait a week or two to make sure it didn't break anything
